airflow/contrib/hooks/imap_hook.py

```diff
@@ -0,0 +1,262 @@
+        """
+        Downloads mail's attachments in the mail folder by its name
+        to the local directory.
+
+        :param name: The name of the attachment that will be downloaded.
+        :type name: str
+        :param local_output_directory: The output directory on the local machine
+                                       where the files will be downloaded to.
+        :type local_output_directory: str
+        :param mail_folder: The mail folder where to look at.
+                            The default value is 'INBOX'.
+        :type mail_folder: str
+        :param check_regex: Checks the name for a regular expression.
+                            The default value is False.
+        :type check_regex: bool
+        :param latest_only: If set to True it will only download
+                            the first matched attachment.
+                            The default value is False.
+        :type latest_only: bool
+        """
+        mail_attachments = self._retrieve_mails_attachments_by_name(name, mail_folder,
+                                                                    check_regex,
+                                                                    latest_only)
+        self._download_files(mail_attachments, local_output_directory)
+
+    def _retrieve_mails_attachments_by_name(self, name, mail_folder, check_regex,
+                                            latest_only):
+        all_matching_attachments = []
+
+        self.mail_client.select(mail_folder)
+
+        for mail_id in self._list_mail_ids_desc():
+            response_mail_body = self._fetch_mail_body(mail_id)
+            matching_attachments = self._check_mail_body(response_mail_body, name,
+                                                         check_regex, latest_only)
+            if matching_attachments:
+                all_matching_attachments.extend(matching_attachments)
+                if latest_only:
+                    break
+
+        self.mail_client.close()
+
+        return all_matching_attachments
+
+    def _list_mail_ids_desc(self):
+        result, data = self.mail_client.search(None, 'All')
+        mail_ids = data[0].split()
+        return reversed(mail_ids)
+
+    def _fetch_mail_body(self, mail_id):
+        result, data = self.mail_client.fetch(mail_id, '(RFC822)')
+        mail_body = data[0][1]  # The mail body is always in this specific location
+        mail_body_str = mail_body.decode('utf-8')
+        return mail_body_str
+
+    def _check_mail_body(self, response_mail_body, name, check_regex, latest_only):
+        mail = Mail(response_mail_body)
+        if mail.has_attachments():
+            return mail.get_attachments_by_name(name, check_regex, find_first=latest_only)
+
+    def _download_files(self, mail_attachments, local_output_directory):
+        for name, payload in mail_attachments:
+            with open(local_output_directory + '/' + name, 'wb') as file:
```

Comment:
What it the attachment is called "../../../../../../etc/passwd"? (Or something else that the airflow user actually has permission to write to.

Reply:
What do you mean? ..whether it would override a file or not?

Reply:
We should put a check in to ensure that the file we create doesn't "escape" the local_output_directory

Also we should consider what happens if the attachment is a symlink (not sure that's possible in mail but...?)
Not this exact case, but this describes a similar problem that could happen https://rt.cpan.org/Public/Bug/Display.html?id=125523

Reply:
Agree. It should check for "escaping" and check for symlinks.
I will implement this.